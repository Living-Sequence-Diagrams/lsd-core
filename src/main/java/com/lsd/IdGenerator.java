package com.lsd;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;

import static java.util.UUID.randomUUID;

public class IdGenerator {
    private final AtomicInteger counter = new AtomicInteger();

    private final Supplier<String> idSupplier;

    public IdGenerator(boolean isDeterministic) {
        if (isDeterministic) {
            idSupplier = () -> String.valueOf(counter.incrementAndGet());
        } else {
            idSupplier = () -> randomUUID().toString()
                    .replaceAll("-", "");
        }
    }

    public String next() {
        return idSupplier.get();
    }

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    public void reset() {
        counter.set(0);
    }
}
