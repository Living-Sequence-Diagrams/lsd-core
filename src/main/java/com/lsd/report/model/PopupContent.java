package com.lsd.report.model;

import j2html.tags.ContainerTag;
import j2html.tags.UnescapedText;

import static j2html.TagCreator.*;

/**
 * Simplifies adding content that is only displayed when a hyperlink is clicked - the id used on the href must match.
 */
public class PopupContent {

    public static ContainerTag popupDiv(String id, String title, String content) {
        return div().withId(id).withClass("overlay").attr("onclick", "location.href='#!';").with(
                div().withClass("popup").attr("onclick", "event.stopPropagation();").with(
                        h2(title),
                        a().withClass("close").withHref("#!").with(new UnescapedText("&times;")),
                        div().withClass("content").with(
                                pre(content)
                        )
                )
        );
    }
}
