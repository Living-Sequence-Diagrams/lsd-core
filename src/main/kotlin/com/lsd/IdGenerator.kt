package com.lsd

import java.util.UUID.randomUUID
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Supplier

class IdGenerator(isDeterministic: Boolean = false) {
    private val counter = AtomicInteger()
    
    private val idSupplier: Supplier<String> = if (isDeterministic)
        Supplier { counter.incrementAndGet().toString() }
    else
        Supplier { randomUUID().toString().replace("-".toRegex(), "") }


    operator fun next(): String = idSupplier.get()

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    fun reset() = counter.set(0)
}