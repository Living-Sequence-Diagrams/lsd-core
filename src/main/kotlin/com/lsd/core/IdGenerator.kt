package com.lsd.core

import java.util.UUID.randomUUID
import java.util.concurrent.atomic.AtomicInteger

private val randomIdSupplier: () -> String = { randomUUID().toString().replace("-".toRegex(), "") }

class IdGenerator(isDeterministic: Boolean = false) {
    private val counter = AtomicInteger()
    private val deterministicIdSupplier: () -> String = { counter.incrementAndGet().toString() }

    private val idSupplier: () -> String = if (isDeterministic) deterministicIdSupplier else randomIdSupplier

    fun next(): String = idSupplier.invoke()

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    fun reset() = counter.set(0)
}