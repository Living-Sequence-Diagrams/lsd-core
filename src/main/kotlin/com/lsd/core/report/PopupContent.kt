package com.lsd.core.report

import com.lsd.core.IdGenerator

object PopupContent {

    private val idGenerator = IdGenerator()

    /**
     * Convenience method to create the html for a hyperlink that is linked to a popup containing the provided content.
     * 
     * @param id The id is random by default and usually won't need to be overridden unless a specific id is required for some reason.
     * @param popupTitle The popupTitle is displayed as the title on the popup box containing the provided content. It's optional.
     * @param hyperlinkText The text that gets displayed on the hyperlink.
     * @param popupContent The content that will be hidden until a user clicks on the hyperlink.
     */
    @JvmStatic
    fun popupHyperlink(
        id: String = idGenerator.next(),
        popupTitle: String = "",
        hyperlinkText: String,
        popupContent: String
    ): String {
        return """
            <a href="#$id">$hyperlinkText</a>
            <div id="$id" class="overlay" onclick="location.href='#!';">
                <div class="popup" onclick="event.stopPropagation();">
                    <h2>$popupTitle</h2>
                    <a class="close" href="#!">&times;</a>
                    <div class="content">$popupContent</div>
                </div>
            </div>
        """.trimMargin()
    }

}