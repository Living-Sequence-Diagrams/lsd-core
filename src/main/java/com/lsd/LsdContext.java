package com.lsd;

import com.lsd.diagram.ComponentDiagramGenerator;
import com.lsd.diagram.SequenceDiagramGenerator;
import com.lsd.events.SequenceEvent;
import com.lsd.parse.MessageParser;
import com.lsd.parse.Parser;
import com.lsd.parse.SynchronousResponseParser;
import com.lsd.properties.LsdProperties;
import com.lsd.report.HtmlIndexWriter;
import com.lsd.report.HtmlReportRenderer;
import com.lsd.report.HtmlReportWriter;
import com.lsd.report.model.DataHolder;
import com.lsd.report.model.Participant;
import com.lsd.report.model.Report;
import com.lsd.report.model.Scenario;

import java.nio.file.Path;
import java.util.*;

import static com.lsd.properties.LsdProperties.DETERMINISTIC_IDS;
import static com.lsd.properties.LsdProperties.MAX_EVENTS_PER_DIAGRAM;
import static java.util.stream.Collectors.toList;

public class LsdContext {

    private static final LsdContext INSTANCE = new LsdContext();

    private final List<CapturedScenario> capturedScenarios = new ArrayList<>();
    private final List<CapturedReport> capturedReports = new ArrayList<>();
    private final List<Participant> participants = new ArrayList<>();
    private final Set<String> includes = new LinkedHashSet<>();
    private final IdGenerator idGenerator = new IdGenerator(LsdProperties.getBoolean(DETERMINISTIC_IDS));
    private final List<Parser> parsers = parsers();
    private final HtmlReportWriter htmlReportWriter = new HtmlReportWriter(new HtmlReportRenderer());
    private final int maxEventsPerDiagram = LsdProperties.getInt(MAX_EVENTS_PER_DIAGRAM);

    private CapturedScenario currentScenario = new CapturedScenario();

    /**
     * Convenience method to create a singleton instance.
     * If you prefer to use other means to share a single instance or
     * want to create multiple instances then use the default constructor instead.
     *
     * @return A singleton instance of LsdContext
     */
    public static LsdContext getInstance() {
        return INSTANCE;
    }

    public void addParticipants(List<Participant> additionalParticipants) {
        participants.addAll(additionalParticipants);
    }

    public void includeFiles(Set<String> additionalIncludes) {
        includes.addAll(additionalIncludes);
    }

    public void addFact(String key, String value) {
        currentScenario.addFact(key, value);
    }

    public void capture(SequenceEvent event) {
        currentScenario.add(event);
    }

    /**
     * Allow string representations of events to be interpreted. If none of the parsers match the pattern then nothing will be captured
     *
     * @param pattern The input string that represents some event to be captured on the sequence diagram
     * @param body    The extra data associated with the event (may be null).
     */
    public void capture(String pattern, String body) {
        var event = parsers.stream()
                .map(parser -> parser.parse(pattern, body))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .findFirst();

        event.ifPresent(this::capture);
    }

    public void completeScenario(String title, String description, OutcomeStatus status) {
        currentScenario.setTitle(title);
        currentScenario.setDescription(description);
        currentScenario.setOutcomeStatus(status);
        capturedScenarios.add(currentScenario);
        currentScenario = new CapturedScenario();
    }

    public Path completeReport(String title) {
        Report report = buildReport(title);
        Path path = htmlReportWriter.writeToFile(report);
        capturedReports.add(new CapturedReport(report.getTitle(), path.getFileName().toString(), report.getStatus()));
        capturedScenarios.clear();
        currentScenario = new CapturedScenario();
        return path;
    }

    public String generateReport(String title) {
        Report report = buildReport(title);
        return htmlReportWriter.writeToString(report);
    }

    public Path createIndex() {
        return HtmlIndexWriter.writeToFile(capturedReports);
    }

    public void clear() {
        idGenerator.reset();
        capturedScenarios.clear();
        capturedReports.clear();
        currentScenario = new CapturedScenario();
    }

    public IdGenerator getIdGenerator() {
        return idGenerator;
    }

    private Report buildReport(String title) {
        return Report.builder()
                .title(title)
                .status(determineOverallStatus(capturedScenarios))
                .scenarios(capturedScenarios.stream()
                        .map(capturedScenario -> Scenario.builder()
                                .title(capturedScenario.getTitle())
                                .id(idGenerator.next())
                                .status(capturedScenario.getOutcomeStatus().getCssClass())
                                .description(capturedScenario.getDescription())
                                .facts(capturedScenario.getFacts())
                                .dataHolders(capturedScenario.getSequenceEvents().stream()
                                        .filter(DataHolder.class::isInstance)
                                        .map(DataHolder.class::cast)
                                        .collect(toList()))
                                .sequenceDiagram(SequenceDiagramGenerator.builder()
                                        .idGenerator(idGenerator)
                                        .events(capturedScenario.getSequenceEvents())
                                        .participants(participants)
                                        .includes(includes)
                                        .build()
                                        .diagram(maxEventsPerDiagram)
                                        .orElse(null))
                                .componentDiagram(ComponentDiagramGenerator.builder()
                                        .idGenerator(idGenerator)
                                        .events(capturedScenario.getSequenceEvents())
                                        .participants(participants)
                                        .build()
                                        .diagram()
                                        .orElse(null))
                                .build())
                        .collect(toList()))
                .build();
    }

    private String determineOverallStatus(List<CapturedScenario> capturedScenarios) {
        return capturedScenarios.stream()
                .map(CapturedScenario::getOutcomeStatus)
                .sorted()
                .findFirst()
                .map(OutcomeStatus::getCssClass)
                .orElse("");
    }

    private List<Parser> parsers() {
        return List.of(
                new MessageParser(idGenerator),
                new SynchronousResponseParser(idGenerator)
        );
    }
}
