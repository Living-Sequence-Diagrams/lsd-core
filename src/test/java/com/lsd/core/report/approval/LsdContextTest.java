package com.lsd.core.report.approval;

import com.lsd.core.IdGenerator;
import com.lsd.core.LsdContext;
import com.lsd.core.domain.*;
import com.lsd.core.report.PopupContent;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.LinkedHashSet;
import java.util.List;

import static com.lsd.core.builders.ActivateLifelineBuilder.activation;
import static com.lsd.core.builders.DeactivateLifelineBuilder.deactivation;
import static com.lsd.core.builders.MessageBuilder.messageBuilder;
import static com.lsd.core.domain.MessageType.*;
import static com.lsd.core.domain.ParticipantType.*;
import static com.lsd.core.domain.Status.*;
import static j2html.TagCreator.*;

class LsdContextTest {

    private final LsdContext lsdContext = LsdContext.getInstance();

    private final IdGenerator idGenerator = lsdContext.getIdGenerator();

    private final LinkedHashSet<String> additionalIncludes = new LinkedHashSet<>(List.of(
            "tupadr3/font-awesome-5/hamburger",
            "tupadr3/font-awesome-5/heart"
    ));

    private final Participant arnie = ACTOR.called("A", "Arnie");
    private final Participant bettie = BOUNDARY.called("Bettie");
    private final Participant cat = CONTROL.called("Cat");
    private final Participant unused = BOUNDARY.called("Unused participant");

    @BeforeEach
    public void clearContext() {
        lsdContext.clear();
        lsdContext.addParticipants(arnie, bettie, cat, unused);
    }

    @Test
    void createsReportWithScenariosAndEvents() {
        lsdContext.includeFiles(additionalIncludes);

        lsdContext.capture(messageBuilder().id(nextId()).to(arnie).label("in").data("start some job").type(SHORT_INBOUND).build());
        lsdContext.capture(messageBuilder().id(nextId()).from(arnie).to(bettie).label("Message 1").data("some data 1").type(BI_DIRECTIONAL).build());
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).label("out").data("some data 1").type(SHORT_OUTBOUND).build());
        lsdContext.capture(messageBuilder().id(nextId()).label("An interaction description that is long enough to need abbreviating").from("Beta").to("Gamma").data("β").type(LOST).build());
        lsdContext.capture(messageBuilder().id(nextId()).label("A synchronous response").from("Gamma").to("Beta").data("200 OK").type(SYNCHRONOUS_RESPONSE).build());
        lsdContext.completeScenario("A Success scenario", "Given a first scenario description<br/>When something happens<br/>Then something else happens", SUCCESS);

        lsdContext.capture(messageBuilder().id(nextId()).label("Sending food <$hamburger{scale=0.4}>").from(arnie).to(bettie).colour("orange").build());
        lsdContext.capture(new TimeDelay(null));
        lsdContext.capture("Sending a response from B to A [#red]", "Thank You!");
        lsdContext.capture(new NoteLeft("Friends <$heart{scale=0.4,color=red}>", null));
        lsdContext.addFact("Something to highlight", "Lorem");
        lsdContext.addFact("Something else to highlight", "amet");
        lsdContext.addFact("Something else to highlight", "consectetur");
        lsdContext.addFact("Something else to highlight", "Thank you!");
        lsdContext.completeScenario("A Warning scenario", p(
                text("A popup with a long text that needs scrolling: "),
                a().withHref("#" + "kljasdlfj").withText("click me!"),
                PopupContent.popupDiv("kljasdlfj", "I am popup", ".. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.")
        ).render(), FAILURE);

        lsdContext.capture("a request from Beta to Gamma", "Please do something");
        lsdContext.capture("sync a synchronous response from Gamma to Beta", "Some Error (123456)");
        lsdContext.addFact("some important value", "123456");
        lsdContext.completeScenario("An Error scenario", "<p>Failure! Expected value to be 123 but was 123456</p>", ERROR);

        Approvals.verify(lsdContext.completeReport("Approval Report").toFile());
    }

    @Test
    void diagramIsSplitOnNewPageEvents() {
        lsdContext.includeFiles(additionalIncludes);

        lsdContext.capture(new PageTitle("A title"));
        lsdContext.capture(messageBuilder().id(nextId()).to(arnie).label("in").data("start some job").type(SHORT_INBOUND).build());
        lsdContext.capture(new Newpage(new PageTitle("A second page")));
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).label("out").data("some data 1").type(SHORT_OUTBOUND).build());
        lsdContext.capture(new Newpage(new PageTitle("A third page")));
        lsdContext.capture(messageBuilder().id(nextId()).label("An interaction description that is long enough to need abbreviating").from("Beta").to("Gamma").data("β").type(LOST).build());
        lsdContext.completeScenario("A Success scenario", "Given a first scenario description<br/>When something happens<br/>Then something else happens", SUCCESS);

        Approvals.verify(lsdContext.completeReport("Split by NewPage Report").toFile());
    }

    @Test
    void notesCanBeAddedToParticipants() {
        lsdContext.capture(new PageTitle("Adding notes"));
        lsdContext.capture(messageBuilder().id(nextId()).from(arnie).to(bettie).label("Good day to you!").build());
        lsdContext.capture(new NoteLeft("Left note", null));
        lsdContext.capture(new NoteRight("Right note", null));
        lsdContext.capture(new TimeDelay("a few seconds later"));
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(cat).label("Good day to you!").build());
        lsdContext.capture(new NoteLeft("Left of Cat", cat.getComponentName()));
        lsdContext.capture(new NoteRight("Right of Cat", cat.getComponentName()));
        lsdContext.capture(new NoteRight("Right note", null));
        lsdContext.capture(new LogicalDivider("a divider"));
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(arnie).label("Me?").build());
        lsdContext.capture(new NoteRight("Right note", null));
        lsdContext.capture(new VerticalSpace(20));
        lsdContext.capture(new NoteLeft("Left of Bettie", bettie.getComponentName()));
        lsdContext.capture(new NoteRight("Right of Bettie", bettie.getComponentName()));

        lsdContext.completeScenario("A Success scenario", "", SUCCESS);

        Approvals.verify(lsdContext.completeReport("Adding notes to participants").toFile());
    }

    @Test
    void participantAreColoured() {
        var greg = ACTOR.called("Greg", null, "green");
        var rosie = BOUNDARY.called("Rosie", null, "red");
        var bart = CONTROL.called("Bart", null, "blue");

        lsdContext.addParticipants(greg, rosie, bart);

        lsdContext.capture(new PageTitle("Adding colours"));
        lsdContext.capture(messageBuilder().id(nextId()).from(greg).to(rosie).label("Roses are red!").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(rosie).to(bart).label("Voilets are blue!").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(bart).to(greg).label("Grass is green?").build());

        lsdContext.completeScenario("A Success scenario", "", SUCCESS);

        Approvals.verify(lsdContext.completeReport("Adding colours to participants").toFile());
    }

    @Test
    void lifelineCanBeActivated() {
        lsdContext.capture(messageBuilder().id(nextId()).from(arnie).to(bettie).label("Good day to you!").build());
        lsdContext.capture(activation().of(bettie).colour("red").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(cat).label("Good day to you!").build());
        lsdContext.capture(deactivation().of(bettie).build());
        lsdContext.capture(activation().of(cat).colour("blue").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(arnie).label("Me?").build());
        lsdContext.capture(deactivation().of(cat).build());

        lsdContext.completeScenario("Lifeline activation/deactivation", null, SUCCESS);

        Approvals.verify(lsdContext.completeReport("Activating lifelines").toFile());
    }

    @Test
    void generateCombinedComponentDiagramSource() {
        var dan = CONTROL.called("Dan");

        lsdContext.addParticipants(arnie, bettie, cat, dan);

        lsdContext.capture(messageBuilder().id(nextId()).from(arnie).to(bettie).label("message1").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(cat).label("message2").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(arnie).label("message3").build());
        lsdContext.completeScenario("Scenario 1", "", SUCCESS);
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(arnie).label("message4").build());
        lsdContext.completeScenario("Scenario 2", "", SUCCESS);
        lsdContext.completeReport("Report 1");

        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(arnie).label("message5").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(bettie).label("message6").build());
        lsdContext.completeScenario("Scenario 3", "", SUCCESS);
        lsdContext.completeReport("Report 2");

        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(dan).label("message5").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(dan).to(arnie).label("message6").build());
        lsdContext.completeScenario("Scenario 4", "", SUCCESS);
        lsdContext.completeReport("Report 3");

        Approvals.verify(lsdContext.completeComponentsReport("Relationships").toFile());
    }

    @Test
    void hasNoDiagramSectionWhenThereAreNoEvents() {
        lsdContext.includeFiles(additionalIncludes);
        lsdContext.addFact("Fact", "no diagrams");
        lsdContext.completeScenario("A scenario without events", "This has no diagrams", SUCCESS);

        Approvals.verify(lsdContext.completeReport("No Diagrams Report").toFile());
    }
    
    @Test
    void participantTypeAndAliasCanBeOverridden() {
        var fred1 = CONTROL.called("Fred");
        var fred2 = DATABASE.called("Fred", "Freddy", "purple");

        lsdContext.addParticipants(List.of(fred1, fred2, bettie));
        lsdContext.capture(messageBuilder().id(nextId()).from("Fred").to(bettie).build());
        lsdContext.completeScenario("Participant is updated", "Freddy is used", SUCCESS);

        Approvals.verify(lsdContext.completeReport("Fred is updated").toFile());
    }

    @Test
    void hasNoDiagramSectionWhenEventsAreCleared() {
        lsdContext.capture(messageBuilder().id(nextId()).to("A").label("in").data("start some job").type(SHORT_INBOUND).build());
        lsdContext.capture(messageBuilder().id(nextId()).to("B").label("in").data("start another job").type(SHORT_INBOUND).build());

        lsdContext.clearScenarioEvents();
        lsdContext.completeScenario("A scenario without events (cleared down)", "This has no diagrams", SUCCESS);

        Approvals.verify(lsdContext.completeReport("No Diagrams Report (cleared down)").toFile());
    }

    @Test
    void createsIndex() {
        lsdContext.completeScenario("Scenario 1", "Error", ERROR);
        lsdContext.completeScenario("Scenario 2", "Warn", FAILURE);
        lsdContext.completeScenario("Scenario 3", "Success", SUCCESS);
        lsdContext.completeReport("First Report (Error)");

        lsdContext.completeScenario("Scenario 1", "Warn", FAILURE);
        lsdContext.completeScenario("Scenario 2", "Success", SUCCESS);
        lsdContext.completeReport("Second Report (Warn)");

        lsdContext.completeScenario("Scenario 1", "Success", SUCCESS);
        lsdContext.completeReport("Third Report");

        lsdContext.completeReport("Fourth Report");

        Approvals.verify(lsdContext.createIndex().toFile());
    }

    @Test
    void generateReport() {
        String result = lsdContext.generateReport("Some title");

        Approvals.verify(result);
    }

    private String nextId() {
        return idGenerator.next();
    }
}