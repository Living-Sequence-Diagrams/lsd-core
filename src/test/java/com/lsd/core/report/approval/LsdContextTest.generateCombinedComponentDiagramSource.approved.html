<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns="http://www.w3.org/1999/html" xml:lang="en" lang="en">
<head>
    <title>LSD</title>
    <link rel="stylesheet" href="style.css">
</head>
<body>
<main role="main">
    <h1 class="logo">LSD</h1>
    <h2>Relationships</h2>
    <div class="contents">
        <section class="sequence diagram svg">
            <h3 class="success">Component Diagram</h3>
            <figure><?xml version="1.0" encoding="us-ascii" standalone="no"?><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentStyleType="text/css" height="299px" preserveAspectRatio="none" style="width:119px;height:299px;" version="1.1" viewBox="0 0 119 299" width="119px" zoomAndPan="magnify"><defs/><g><!--entity Arnie--><g id="elem_Arnie"><ellipse cx="48.4902" cy="26" fill="#FFFFFF" rx="8" ry="8" style="stroke:#000000;stroke-width:1.0;"/><path d="M48.4902,34 L48.4902,61 M35.4902,42 L61.4902,42 M48.4902,61 L35.4902,76 M48.4902,61 L61.4902,76 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="37" x="29.9902" y="92.0752">Arnie</text></g><!--entity Bettie--><g id="elem_Bettie"><path d="M27.9902,135 L27.9902,159 M27.9902,147 L44.9902,147 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><ellipse cx="56.9902" cy="147" fill="#FFFFFF" rx="12" ry="12" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="42" x="27.4902" y="177.0752">Bettie</text></g><!--entity Cat--><g id="elem_Cat"><ellipse cx="48.4902" cy="246.5" fill="#FFFFFF" rx="12" ry="12" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="44.4902,234.5,50.4902,229.5,48.4902,234.5,50.4902,239.5,44.4902,234.5" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="24" x="36.4902" y="276.5752">Cat</text></g><!--entity Dan--><g id="elem_Dan"><ellipse cx="98.4902" cy="225" fill="#FFFFFF" rx="8" ry="8" style="stroke:#000000;stroke-width:1.0;"/><path d="M98.4902,233 L98.4902,260 M85.4902,241 L111.4902,241 M98.4902,260 L85.4902,275 M98.4902,260 L111.4902,275 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="28" x="84.4902" y="291.0752">Dan</text></g><path d="M41.8812,95.2547 C41.5526,107.3118 41.7983,120.2681 42.6183,130.9805 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="42.6183,130.9805,45.9197,121.7015,42.2367,125.9951,37.9431,122.312,42.6183,130.9805" style="stroke:#000000;stroke-width:1.0;"/><path d="M42.612,180.102 C41.4527,195.3187 41.4496,215.037 42.6028,230.2774 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="42.6028,230.2774,45.9123,221.0012,42.2256,225.2916,37.9352,221.6048,42.6028,230.2774" style="stroke:#000000;stroke-width:1.0;"/><path d="M34.3905,230.2593 C26.8573,216.1791 18.3573,197.6791 14.4902,180 C9.8366,158.7252 9.8366,152.2748 14.4902,131 C17.602,116.7738 23.7137,102.0161 29.9076,89.4575 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="29.9076,89.4575,22.3392,95.7599,27.696,93.9418,29.5141,99.2985,29.9076,89.4575" style="stroke:#000000;stroke-width:1.0;"/><path d="M54.3621,130.9805 C55.1821,120.2681 55.4278,107.3118 55.0992,95.2547 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="55.0992,95.2547,51.3459,104.3604,55.2354,100.2529,59.3429,104.1424,55.0992,95.2547" style="stroke:#000000;stroke-width:1.0;"/><path d="M54.3776,230.2774 C55.5308,215.037 55.5277,195.3187 54.3684,180.102 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="54.3684,180.102,51.0637,189.3799,54.7483,185.0876,59.0406,188.7722,54.3684,180.102" style="stroke:#000000;stroke-width:1.0;"/><path d="M60.5938,180.102 C67.7614,194.0789 76.8766,211.8535 84.371,226.4676 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="84.371,226.4676,83.8235,216.634,82.0894,222.0185,76.7049,220.2845,84.371,226.4676" style="stroke:#000000;stroke-width:1.0;"/><path d="M95.9964,215.7447 C93.7146,191.1727 89.429,158.7743 81.4902,131 C77.8562,118.2864 72.4417,104.9322 67.0441,93.1219 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="67.0441,93.1219,67.1471,102.9702,69.1225,97.6695,74.4232,99.6449,67.0441,93.1219" style="stroke:#000000;stroke-width:1.0;"/></g></svg></figure>
            <div onclick="location.href='#!';" id="source" class="overlay">
                <div class="popup" onclick="event.stopPropagation();">
                    <h2>PlantUml markup</h2>
                    <span class="copy" title="copy" onclick="copyToClipboard('uml_code')"></span>
                    <a class="close" href="#!">&times;</a>
                    <div class="content">
                        <pre><code id="uml_code">@startuml
!pragma layout smetana
!theme plain
skinparam backgroundColor transparent

actor Arnie
boundary Bettie
control Cat


[Arnie] --&gt; Bettie
[Bettie] --&gt; Cat
[Cat] --&gt; Arnie
[Bettie] --&gt; Arnie
[Cat] --&gt; Bettie
[Bettie] --&gt; Dan
[Dan] --&gt; Arnie


remove @unlinked
@enduml</code></pre>
                    </div>
                </div>
            </div>
            <sub><a href="#source">[source]</a></sub>
        </section>
    </div>
</main>
</body>
</html>